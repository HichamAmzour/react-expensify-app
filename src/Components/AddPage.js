import React from 'react';
import ExpenseForm from './expenseForm';
import {connect} from 'react-redux';
import {addExpenseAction} from '../ReduxFolder/actions/expenseActions';


export class AddExpensePage extends React.Component{

    onSubmit = (expense)=>{
        // props.dispatch(addExpenseAction(expense));
        this.props.addExpense(expense)
        this.props.history.push('/');
    }
    render(){
        return (
            <div>
                <p>this is Add Expense Page component</p>
                <ExpenseForm onSubmit={this.onSubmit}/>
            </div>
        ) 
    }
}

const matchDispatchToProps = (dispatch)=>(
    {
        addExpense : (expense)=>{
            dispatch(addExpenseAction(expense))
        }
    } 
)

export default connect(undefined,matchDispatchToProps)(AddExpensePage);