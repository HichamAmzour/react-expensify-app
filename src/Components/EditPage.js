import React from 'react';
import { connect } from 'react-redux';
import { editExpenseAction, removeExpenseAction } from '../ReduxFolder/actions/expenseActions';
import ExpenseForm from './expenseForm';

export class EditExpensePage extends React.Component {

    editExpense = (expense)=>{
        this.props.EditExpense(this.props.expense.id, expense);
        this.props.history.push('/');
    }

    removeExpense =  (id) => { 
        this.props.removeExpense(id);
        this.props.history.push('/');
    }

    render(){

        return (
            <div>
               <p>Editing the element that has Id: {this.props.match.params.id}</p>
               <ExpenseForm 
                 expense={this.props.expense} 
                 onSubmit={this.editExpense}
                />
            <button onClick={this.removeExpense(this.props.expense.id)} 
            >Remove</button>
            </div>
        )
    }
}

const mapDispatchToProps = (dispatch, props)=>{
    return ({
        EditExpense : (id, expense)=>{
            dispatch(editExpenseAction(id,expense))
        },
        removeExpense : (id)=>{
            dispatch(removeExpenseAction(id))
        }
    })
}

const mapStateToProps= (state, props) => {
  
    return {
        expense: state.expenses.find((expense) => expense.id === props.match.params.id)
    };
};

export default connect(mapStateToProps,mapDispatchToProps)(EditExpensePage);