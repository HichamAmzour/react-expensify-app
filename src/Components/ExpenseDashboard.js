import React from 'react';
import ExpenseList from './ExpensesList';
import ExpenseListFilters from './expenseListFilter';

const ExpenseDashboard=()=>(
    <div>
       <p>this is Expense Dashboard component</p>
       <ExpenseListFilters/>
       <ExpenseList/>
    </div>
);

export default ExpenseDashboard;