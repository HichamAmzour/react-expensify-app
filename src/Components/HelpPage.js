import React from 'react';

const HelpPage=()=>(
    <div>
        <p>this is Help Expense Page component</p>
    </div>
);

export default HelpPage;