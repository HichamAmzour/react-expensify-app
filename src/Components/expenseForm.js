import React from 'react';
import 'react-dates/initialize';
import moment from 'moment';
import { DateRangePicker, SingleDatePicker, DayPickerRangeController } from 'react-dates';
// import 'react-dates/lib/css/_datepicker.css'; // use it in the app main file


// const now=moment();
// console.log(now.format('MMMM Do, YYYY'));
export default class ExpenseForm extends React.Component{

    constructor(props){
        super(props);
        this.state={
            description: props.expense ? props.expense.description : '',
            amount: props.expense ? props.expense.amount/100 : '',
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            notes: props.expense ? props.expense.notes : '',
            calandarFocused:false,
            error:'',
        };
    }

    onAmountChange = (e)=>{
    //    e.presest();
        const amount = e.target.value;
        if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)){
            this.setState(()=>({
                amount
            }));
        }
    };
    onNotesChange = (e)=>{
        const notes=e.target.value;
        this.setState(()=>({
           notes
        }));
    };
    onCreatedAtChange = (e)=>{
        const createdAt=e.target.value;
        this.setState(()=>({
            createdAt
        }));
    };
    onDescriptionChange = (e)=>{
        const description= e.target.value
        this.setState(()=>({
            description
        }));
    };

    onDateChange = (createdAt)=>{
        if(createdAt){
            this.setState(()=>({
                createdAt
            }));
        }
    };
    onFocusChange = ({focused}) =>{
        this.setState(()=>({
            calandarFocused: focused
        }));
    };

    setError = (error)=>{
        this.setState(()=>({
            error
        }))
    };

    onSubmit = (e)=>{
        e.preventDefault();
       
        if(!this.state.description || !this.state.amount){
            this.setError("You must provide a descripton/ amount");
        }
        else{
            this.setError("");
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount,10) * 100,
                createdAt: this.state.createdAt.valueOf(),
                notes:this.state.notes
            });
        }
    };

    render(){
        return (
          <div>
            {this.state.error && <p>{this.state.error}</p>}
            <form onSubmit={this.onSubmit}> 
              <label>Description : </label>
              <input value={this.state.description} onChange={this.onDescriptionChange}  /><br/><br/>
              <label>Amount : </label>
              <input value={this.state.amount} onChange={this.onAmountChange}/><br/><br/>
              <label type="number">Notes : </label>
              <textarea  value={this.state.notes} onChange={this.onNotesChange}  /><br/><br/>
              <div>
              <label type="number">CreatedAt : </label>
                <SingleDatePicker
                    date={this.state.createdAt}
                    onDateChange={this.onDateChange}
                    focused={this.state.calandarFocused}
                    onFocusChange={this.onFocusChange}
                    numberOfMonths={1}
                    isOutsideRange={()=> false}
                />
              </div>
              {
                  !this.props.expense ?  <button>Add Expense</button> : <button>Edit Expense</button>
              }
              
            </form>
          </div>
        );
    }

}