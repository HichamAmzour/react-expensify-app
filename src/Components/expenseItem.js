import React from 'react';
import {Link} from 'react-router-dom';


const ExpenesItem = ( {description, notes, amount, createdAt, id} )=>(
    <div key={id}>
       <hr/>
        <h4 >Description : {description}</h4>
        <h4 >Amount      : {amount}</h4>
        <h4 >Notes       : {notes}</h4>
        <h4 >Created At  : {createdAt}</h4>
        <Link to={'/edit/'+id}>Edit</Link>
        <hr/>
    </div>
);

export default ExpenesItem;

