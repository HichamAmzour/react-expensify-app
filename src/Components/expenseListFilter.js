import React from 'react';
import { connect} from 'react-redux';
import { 
    setTextFiltersAction,
    sortByAmountAction,
    sortByDateAction,
    setStartDateAction,
    setEndDateAction
} from '../ReduxFolder/actions/filterActions';

import { DateRangePicker } from 'react-dates';

class ExpenseListFilters extends React.Component{

    state={
        focusedInput:null
    }
                    //its takes an object
    onDatesChange = ({startDate, endDate})=>{
        this.props.dispatch(setStartDateAction(startDate));
        this.props.dispatch(setEndDateAction(endDate));
    };

    onFocusChange = (focusedInput)=>{
        this.setState(()=>({focusedInput}));
    };

    render(){
        return (
            <div>
                <input value={this.props.filters.text} onChange={(e)=>{
                    this.props.dispatch(setTextFiltersAction(e.target.value));
                }}/>
                <div>
                    <DateRangePicker
                        startDate={this.props.filters.startDate}
                        startDateId={this.props.filters.startDate+""}
                        endDate={this.props.filters.endDate}
                        endDateId={this.props.filters.endDate+""}
                        onDatesChange={this.onDatesChange}
                        onFocusChange={this.onFocusChange}
                        focusedInput={this.state.focusedInput}
                        numberOfMonths={1}
                        isOutsideRange={()=> false}
                        showClearDates={true}
                    />
                </div>
                <select value={this.props.filters.sortBy} onChange={(e)=>{
                    e.target.value ==='amount' ? 
                    this.props.dispatch(sortByAmountAction()):
                    this.props.dispatch(sortByDateAction())
                }}>
                    <option value="date">By Date</option>
                    <option value="amount">By Amount</option>
                </select>
            </div>
        );
    }

}


export default connect((state)=>({
    filters : state.filters
}))(ExpenseListFilters);

