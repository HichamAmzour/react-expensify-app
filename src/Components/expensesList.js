import React from 'react';
import {connect} from 'react-redux';
import ExpenseItem from './expenseItem';
import getVisibleExpenses from '../ReduxFolder/selectors/expensesSelector';

//casue we will this this component not connected component to the store
export const ExpenseList = (props)=>(
    <div>
    <h1>Expense List</h1>
    {
        props.expenses.length === 0 ? (<p>There is no expenses in the list</p>) :
        props.expenses.map((ex)=>(
            <ExpenseItem key={ex.id} {...ex}/>
        ))
    }
    </div>
);
const mapStateToProps = (state)=>{
    return {
        expenses : getVisibleExpenses(state.expenses,state.filters)
    };
}
const ConnectToStore = connect(mapStateToProps)(ExpenseList);
export default ConnectToStore;