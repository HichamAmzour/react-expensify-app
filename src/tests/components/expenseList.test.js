import React from 'react';
import {shallow} from 'enzyme';
import {ExpenseList} from '../../Components/expensesList';
import expensesList from '../fixturesDb/expensesList';

test("Should render expense list with expenses items",()=>{
    const wrapper =  shallow(<ExpenseList expenses={expensesList}/>);
    expect(wrapper).toMatchSnapshot();
});

test("Should render a message when the expenses list is Empety ",()=>{
    const wrapper =  shallow(<ExpenseList expenses={[]}/>);
    expect(wrapper).toMatchSnapshot();
});
