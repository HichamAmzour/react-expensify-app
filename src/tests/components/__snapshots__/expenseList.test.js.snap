// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render a message when the expenses list is Empety  1`] = `
<div>
  <h1>
    Expense List
  </h1>
  <p>
    There is no expenses in the list
  </p>
</div>
`;

exports[`Should render expense list with expenses items 1`] = `
<div>
  <h1>
    Expense List
  </h1>
  <ExpenesItem
    amount={150}
    createdAt={-345600000}
    description="rent"
    id="1"
    key="1"
    notes=""
  />
  <ExpenesItem
    amount={50}
    createdAt={345600000}
    description="wifi bill"
    id="2"
    key="2"
    notes=""
  />
  <ExpenesItem
    amount={250}
    createdAt={864000000}
    description="water bill"
    id="3"
    key="3"
    notes=""
  />
</div>
`;
